Mathlib/SetTheory/Ordinal/Clubs.lean

```diff
@@ -0,0 +1,243 @@
+    (hSunb T hT) (pos_of_gt hp).ne.symm
+  let f : ↑S → Ordinal := fun U ↦ lift.{u + 1, u} (sInf (U ∩ (Ioo p o)))
+  have infMem : ∀ U : S, sInf (↑U ∩ Ioo p o) ∈ ↑U ∩ Ioo p o := fun U ↦
+    csInf_mem (inter_suffix_of_isUnbounded (hSunb U.1 U.2) hp : (↑U ∩ Ioo p o).Nonempty)
+  have flto : ∀ U : S, f U < lift.{u + 1, u} o := fun U ↦ by
+    simp_all only [mem_inter_iff, mem_Ioo, lift_lt, f]
+  set q := (Ordinal.sup.{u + 1, u} f) + 1 with qdef
+  have qlto : q < lift.{u + 1, u} o :=
+    ((lift_isLimit.{u + 1, u} o).mpr oLim).2 (sup.{u + 1, u} f) (sup_lt_ord hScard flto)
+  rcases lift_down qlto.le with ⟨q', hq'⟩
+  use q'
+  have fltq : ∀ U, f U < q := fun U ↦
+    lt_of_le_of_lt (le_sup.{u + 1, u} f U) (qdef ▸ lt_add_one (sup f))
+  constructor <;> try constructor
+  · exact lift_lt.mp (hq' ▸ qlto)
+  · rcases hSemp with ⟨U, hU⟩
+    have pltf : lift.{u + 1, u} p < f ⟨U, hU⟩ :=
+      lift_lt.mpr (mem_of_mem_inter_right (infMem ⟨U, hU⟩)).1
+    have := lt_of_lt_of_le pltf (fltq ⟨U, hU⟩).le
+    rwa [← hq', lift_lt] at this
+  intro U hU
+  specialize infMem ⟨U, hU⟩
+  specialize fltq ⟨U, hU⟩
+  have : f ⟨U, hU⟩ ∈ Ioo (lift.{u + 1, u} p) q := ⟨lift_lt.mpr infMem.2.1, fltq⟩
+  rw [← hq'] at fltq
+  rcases lift_down fltq.le with ⟨fUdown, fUlift⟩
+  use fUdown
+  constructor
+  · simp_all only [lift_inj, mem_inter_iff, f]
+  · constructor
+    exact lift_lt.mp <| fUlift ▸ (this.1)
+    exact lift_lt.mp <| hq' ▸ (fUlift ▸ this).2
+
+theorem monotone_of_succ_lt_omega (f : Π p < ω, Iio o)
+    (hf : ∀ i, (hi : i < ω) → f i hi < f (i + 1) (omega_isLimit.2 i hi)) :
+    ∀ (i j hi hj), i < j → f i hi < f j hj := fun i j hi hj iltj ↦ by
+  have : (hj : j < ω) → ∀ k, (kltj : k < j) → f k (kltj.trans hj) < f j hj := Ordinal.limitRecOn j
+    (by
+      intro _ _ h
+      cases (Ordinal.zero_le _).not_lt h)
+    (by
+      intro i ih hj k hk
+      cases (lt_succ_iff.mp hk).lt_or_eq with
+      | inl h =>
+          specialize ih ((lt_succ i).trans hj) k h
+          exact ih.trans <| hf i ((lt_succ i).trans hj)
+      | inr h =>
+          simp_rw [h]
+          exact hf i ((lt_succ i).trans hj))
+    (by
+      intro _ h _ h'
+      cases (omega_le_of_isLimit h).not_lt h')
+  exact this hj i iltj
+
+/--
+Given a limit ordinal `o` and a property on pairs of ordinals `P`, such that
+for any `p < o` there is a `q < o` above `p` so that `P p q`, we can construct
+an increasing `ω`-sequence below `o` that satisfies `P` between every 2 consecutive elements.
+Additionaly, the sequence can begin arbitrarily high in `o`. That is, above any `r < o`.
+-/
+theorem exists_omega_seq_succ_prop (opos : 0 < o) {P : Ordinal → Ordinal → Prop}
+    (hP : ∀ p < o, ∃ q < o, (p < q ∧ P p q)) {r} (rlto : r < o) : ∃ f : Π p < ω, (Iio o),
+    (∀ i : Ordinal.{u}, (hi : i < ω) → P (f i hi) (f (i + 1) (omega_isLimit.2 i hi)))
+    ∧ (∀ i j, (hi : i < ω) → (hj : j < ω) → (i < j) → f i hi < f j hj)
+    ∧ r < f 0 omega_pos := by
```

Comment:
I used `ω` because this is supposed to be used with theorems like `isAcc_bsup_of_between` that work with arbitrary limit ordinals, doing it with `ℕ` first is mostly an implementation detail.

Reply:
[Ordinal.lift_type_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/Basic.html#Ordinal.lift_type_eq) or Ordinal.type_eq and (the definition) [Ordinal.type_nat_lt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/Basic.html#Ordinal.type_nat_lt) will be useful to construct a RelIso between `ℕ` and `Iio ω`. To compute the `type` of `Iio o` we may extract from [the proof of `small_Iio`](https://github.com/leanprover-community/mathlib4/blob/04e23488cdc6c40e3480f98100698e0566c1c192/Mathlib/SetTheory/Ordinal/Arithmetic.lean#L1169-L1178).

Reply:
I defined the RelIso, since this construction is already done I'll keep it as it is, but we have it for the future.